docs/using-the-compiler.rst

```diff
@@ -0,0 +1,141 @@
+In order to ease source code verification of complex contracts that are spread across several files,
+there is a standardized for describing the relations between those files.
+Furthermore, the compiler can generate a json file while compiling that includes
+the source, natspec comments and other metadata whose hash is included in the
+actual bytecode.
+
+There is some overlap between the input description and the metadata output
+and due to the fact that some fields are optional, the metadata can be used as
+input to the compiler. In order to verify the metadata, you actually take it,
+re-run the compiler on the metadata and check that it again produces the same
+metadata.
+
+If the compiler is invoked in a different way, not using the input
+description (for example by using a file content retrieval callback),
+the compiler can still generate the metadata alongside the bytecode of each
+contract.
+
+The metadata standard is versioned. Future versions are only required to provide the "version" field,
+the two keys inside the "compiler" field. The field compiler.keccak should be the keccak hash
+of a binary of the compiler with the given version.
+
+The example below is presented in a human-readable way. Properly formatted metadata
+should use quotes correctly, reduce whitespace to a minimum and sort the keys of all objects
+to arrive at a unique formatting.
+
+Comments are of course not permitted and used here only for explanatory purposes.
+
+Input Description
+-----------------
+
+The input description could change with each compiler version, but it
+should be backwards compatible if possible.
+
+    {
+      sources:
+      {
+        "abc": "contract b{}",
+        "def": {keccak: "0x123..."}, // source has to be retrieved by its hash
+        "dir/file.sol": "contract a {}"
+      },
+      settings:
+      {
+        remappings: [":g/dir"],
+        optimizer: {enabled: true, runs: 500},
+        compilationTarget: "myFile.sol:MyContract", // Can also be an array
+        // To be backwards compatible, use the full name of the contract in the output
+        // only if there are name clashes.
+        // If the full name was given as "compilationTargets", use the full name.
+        libraries: {
+          "def:MyLib": "0x123123..."
+        },
+        // The following can be used to restrict the fields the compiler will output.
+        outputSelection: {
+          // to be defined
+        }
+      }
+    }
+
+Metadata Output
+---------------
+
+Note that the actual bytecode is not part of the metadata because the hash
+of the metadata structure will be included in the bytecode itself.
+
+This requires the compiler to be able to compute the hash of its own binary,
+which requires it to be statically linked. The hash of the binary is not
+too important. It is much more important to have the commit hash because
+that can be used to query a location of the binary (and whether the version is
+"official") at a registry contract. 
+
+    {
```

Comment:
I actually did not think of this as a generic thing, but it might very well be :-)
Added language field.